### DISCUSSION TITLE
Get information from the browser

### DISCUSSION THREAD

**Discussion by thnewlands** [Voltstro-Studios/UnityWebBrowser#71] [Requests]
Hi! Is there some way to get the scroll position from the browser? If not it would be interesting to be able to retrieve basic information from the browser.

**Comment by thnewlands** [comment]
Experimenting with ExecuteJs right now. This project is amazing! Still need to figure out how to read results back.

**Reply by thnewlands** [reply]
I was able to solve this by reading the console log -- I think the best way to handle this would probably be to use websockets w/ the ports?

**Reply by thnewlands** [reply]
Did more research and it seems like the best way to go would be to use ZeroMQ

**Reply by thnewlands** [reply]
I think my idea is to be able to receive ZFrame data through callbacks in at the level of the WebBrowserUI.cs script.

**Reply by thnewlands** [reply]
Open to other suggestions.

**Reply by thnewlands** [reply]
I guess I could listen to a port and use https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/postMessage & https://docs.unity3d.com/2020.1/Documentation/Manual/UNetReaderWriter.html

**Reply by Voltstro** [reply]
Since you are talking about ZeroMQ and WebBrowserUI, I am going to assume you are using UWB 1.6.3.
While still marked as preview, I would recommend using 2.0.0 version instead, since it uses our VoltRpc for handling communications, and with VoltRpc, its really easy to add to the interface for getting data (such as getting the scroll position).

**Reply by thnewlands** [reply]
That's rad! I'll have to give that a shot. Thank you. Didn't mean to barrage the maintainers of this project with emails -- just wanted to keep a log of approaches in case anyone else came across this with similar questions.

**Reply by Voltstro** [reply]
I am all good with the questions & logs, write as much as you want.

I had a look at getting the mouse scroll position implemented, you can get it now using `GetScrollPosition()` in the `WebBrowserClient` class.

**Reply by thnewlands** [reply]
Thank you!!

**Reply by thnewlands** [reply]
Just noticed you've got a resize method in v2 too!

**Comment by thnewlands** [comment]
Follow up idea -- is it possible to route CEF audio through an audiosource? It looks like there is a function called CEFAudioHandler that might be able to do something? https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=17691

**Reply by Voltstro** [reply]
Yes, you should be able to do that.
CefAudioHandler exposes OnAudioStreamStarted, OnAudioStreamPacket and OnAudioStreamStopped.

You would need to create a stream of some sort (such as a network stream using pipes) for passing the PCM data back to Unity when OnAudioStreamStarted is invoked, write to that stream when OnAudioStreamPacket is invoked using the provided data, then close it when OnAudioStreamStopped is invoked.

But I am not too sure if an Unity Audio Source can just directly play a PCM stream, I had quick look and it doesn't seem like there is too much info on it, you would have to look into it yourself.

**Reply by thnewlands** [reply]
I'll look into it! If I figure anything out I might make a pull request : )

**Reply by thnewlands** [reply]
Something like this I think would be the way to go: https://gist.github.com/haydenjameslee/4b405cf5e08d227de318

**Reply by thnewlands** [reply]
The key piece is https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnAudioFilterRead.html

**Reply by Voltstro** [reply]
Happy to take any PR related to it.
